tests/test_decoder.py

```diff
@@ -0,0 +1,482 @@
+                    expected_output = 'utf_8'
+                elif not encoding_detectable(string, codec):
+                    expected_output = None
+                elif is_test_codec(codec):
+                    expected_output = test_codec_built_in_equivalent(codec)
+                else:
+                    expected_output = codec
+                self.detect_encoding_test_helper(
+                    string,
+                    string.encode(codec),
+                    expected_output
+                )
+
+    def auto_decode_test_helper(
+        self,
+        input_bytes,
+        codec_for_input_bytes,
+        expected_output
+    ):
+        ERROR = "auto_decode({}) returned the wrong value."
+        does_auto_detect_encodings_return_value_matter = (
+            codec_for_input_bytes is not None and (
+                encoding_detectable(expected_output, codec_for_input_bytes)
+                or len(input_bytes) == 0
+            )
+        )
+        if does_auto_detect_encodings_return_value_matter:
+            actual_output = decoder.auto_decode(input_bytes)
+            self.assertEqual(
+                expected_output,
+                actual_output,
+                msg=ERROR.format(repr(input_bytes))
+            )
+            self.assertIsInstance(actual_output, str)
+        else:
+            try:
+                decoder.auto_decode(input_bytes)
+            except UnicodeDecodeError as exception:
+                return exception
+        return None
+
+    def test_auto_decode_with_pre_encoded_strings(self):
+        ERROR = "auto_decode({}) should not have raised an exception"
+        for pre_encoded_test_string_info in PRE_ENCODED_TEST_STRING_INFOS:
+            exception = self.auto_decode_test_helper(
+                pre_encoded_test_string_info.input_bytes,
+                pre_encoded_test_string_info.codec_for_input_bytes,
+                pre_encoded_test_string_info.expected_output_str
+            )
+            if exception is not None:
+                new_exception = self.failureException(
+                    msg=ERROR.format(
+                        repr(pre_encoded_test_string_info.input_bytes)
+                    )
+                )
+                raise new_exception from exception
+
+    def test_auto_decode_with_strings_encoded_at_runtime(self):
+        at_least_one_decode_error = False
+        for string in TEST_STRINGS_TO_ENCODE_AT_RUNTIME:
+            for codec in UTF_CODECS:
+                exception = self.auto_decode_test_helper(
+                    string.encode(codec),
+                    codec,
+                    string
+                )
+                if exception is not None:
+                    at_least_one_decode_error = True
+        self.assertTrue(
+            at_least_one_decode_error,
+            msg=(
+                "None of the TEST_STRINGS_TO_ENCODE_AT_RUNTIME triggered a "
+                + "decoding error."
+            )
```

Comment:
The `+` is not needed:

```suggestion
            msg=("None of the TEST_STRINGS_TO_ENCODE_AT_RUNTIME triggered a "
                 "decoding error.")
```

Or if you really like Black :wink::
```suggestion
            msg=(
                "None of the TEST_STRINGS_TO_ENCODE_AT_RUNTIME triggered a "
                "decoding error."
            )
```

Reply:
Done.